Fixes #7462 - new UI for network interfaces

tstrachota opened the pull request:
Editing network interfaces moved into a modal window. Some screenshots of the new UI are at:
https://tstrachota.fedorapeople.org/screenshots/nics_ui/

Bits of host form related to NICs are initially hidden and copied into a modal window on edit button click. This allows reusing validation error notices rendered by rails.

Commit:
Fixes #7462 - new UI for network interfaces

tstrachota commented:
@ares @GregSutcliffe

ohadlevy commented on app/views/nic/_hidden_layout.html.erb:
```diff
@@ -0,0 +1,7 @@
+
+<div style="display: none;" id="interfaceHidden<%= local_assigns[:id] %>" interface-id="<%= local_assigns[:id] %>">
```
nitpick: use class=hide

ares commented on app/assets/javascripts/host_edit_interfaces.js:
```diff
@@ -0,0 +1,108 @@
+  row.find('.mac').html(modal_window.find('.interface_mac').val());
+}
+
+
```
nitpick: few extra lines

ares commented on app/assets/javascripts/host_edit_interfaces.js:
```diff
@@ -0,0 +1,108 @@
+  var modal_window = $('#interfaceModal');
+  var interface_id = modal_content.attr('interface-id');
+
+  modal_window.attr('current-id', interface_id);
```
I think using data attributes would be better for current-id

ares commented:
Could you please add some space between green button "Add Interface" and horizontal line below?

ares commented:
I'm not sure whether this issue is introduced by this PR but when I try to add new interface to existing host and fill in only identifier, it allows me to save the host, I see error but the interface is not created. Same applies to new host.

tstrachota commented:
Updated:
- spacing around `+ Add interface` button
- link up/down status icons
- background of invalid interface's row is now light red
- red title and message when there's a validation error in one of the interfaces
- changed class of interface edit button

Screenshots updated too.

tstrachota commented:
@ares the issue you came across with is an error somewhere in validations (it's in development branch too). NICs with blank mac addresses are just removed.

ares commented:
@tstrachota definitely much nicer! Last thing I doubt is the warning icon in edit button (sorry for potential bad hint, I didn't image varying width of button), let's see what others think. Don't worry about the validation issue then, validations may change later and it's not part of this PR.

ares commented:
@tstrachota you made rubocop angry http://ci.theforeman.org/job/test_develop_pr_rubocop/1054/checkstyleResult/source.0/#332 :-)

tstrachota commented:
@ares I'm not fond of the varying size too. We can remove the icon. No strong opinion from my side here.

ohadlevy commented:
IMHO the background color on error is too much. Can we just mark the tr border red (similar to other errors in input fields)?

I think the icon next to edit on error is not needed.maybe change the btn class instead of an icon for f its needed?

tstrachota commented:
@ohadlevy https://tstrachota.fedorapeople.org/screenshots/nics_ui/005_nics_validation_error_v3.png
Better?

ohadlevy commented:
yes, I wonder if we should have the warning (above the table) wrapped in a
warning box?

tstrachota commented:
@ohadlevy updated https://tstrachota.fedorapeople.org/screenshots/nics_ui/006_nics_validation_error_v4.png

ohadlevy commented:
looks better!, does the buttons look better when you pull-right them?
+1 from my side.

tstrachota commented:
Buttons on right seem too far from the rest of the content to me.

ares commented on app/views/hosts/_interfaces.html.erb:
```diff
@@ -0,0 +1,86 @@
+
+<% @interfaces_invalid = @host.interfaces.any?{ |i| !i.valid? } %>
```
nitpick: could you please use interfaces_invalid without @ so we don't risk any conflict in future?

ares commented on app/views/nic/new.js.erb:
```diff
@@ -1,8 +1,8 @@
   <%= fields_for :host, Host::Base.new do |host| %>
     <%= host.fields_for :interfaces, @interface, :child_index => @key do |interfaces| %>
-      $("#host_interfaces_attributes_<%= @key %>_type").closest('fieldset').html(
-        '<%=j render :partial => @interface, :layout => 'nic/interface_layout', :locals => {:f => interfaces} %>'
```
it seems that we don't use nic/interface_layout partial any more so maybe we should remove it?

ares commented:
@tstrachota I'm happy with the code, could you please squash before I'll reapply and test it?

tstrachota commented:
@ares done

ares commented on app/views/hosts/_interfaces.html.erb:
```diff
@@ -0,0 +1,86 @@
+      <tr class="<%= 'has-error' unless interface.valid? %>" id="interface<%= interface.object_id %>" data-interface-id="<%= interface.object_id %>">
+        <td class="status hidden-xs" align="center"><%= link_status(interface) %></td>
+        <td class="identifier ellipsis"><%= interface.identifier %></td>
+        <td class="type hidden-xs"><%= interface.class.humanized_name %></td>
```
missing N_()

ares commented:
Merged as https://github.com/theforeman/foreman/commit/3154d77e023e84a2515c6a9bdcaf974fcdbb4493, thanks @tstrachota!
